contracts/token/ERC721/ERC721Enumerable.sol

```diff
@@ -110,6 +110,35 @@ contract ERC721Enumerable is ERC165, ERC721, IERC721Enumerable {
         _ownedTokensIndex[lastToken] = tokenIndex;
     }
 
+    /**
+     * @dev Internal function to transfer a token ID from the list of a given address to another one.
+     * This function is internal due to language limitations, see the note in ERC721.sol.
+     * It is not intended to be called by custom derived contracts: in particular, it emits no Transfer event,
+     * and doesn't clear approvals.
+     * @param from address representing the previous owner of the given token ID
+     * @param to address representing the new owner of the given token ID
+     * @param tokenId uint256 ID of the token to be transferred from the tokens list of the given address
+     */
+    function _transferToken(address from, address to, uint256 tokenId) internal {
+        super._transferToken(from, to, tokenId);
+
+        // To prevent a gap in from's array, we store from's last token in the index of the token to delete, and
+        // then delete the last slot (swap and pop).
+        uint256 lastTokenIndex = _ownedTokens[from].length.sub(1);
+        uint256 lastTokenId = _ownedTokens[from][lastTokenIndex];
+
+        uint256 preTokenIndex = _ownedTokensIndex[tokenId];
+        _ownedTokens[from][preTokenIndex] = lastTokenId;
+        _ownedTokensIndex[lastTokenId] = preTokenIndex;
+
+        // This also deletes the contents at the last position of the array
+        _ownedTokens[from].length--;
+
+        // Now we add the token to the recipient
+        _ownedTokens[to].push(tokenId);
+        _ownedTokensIndex[tokenId] = _ownedTokens[to].length.sub(1);
```

Comment:
Doesn't the push command return the new length. There's a pattern to do:

`uint256 index = arr.push(val).sub(1);`

Would that be more efficient?

Reply:
Nice! I'm not sure if it will be more efficient than taking the length before the push. I'm guessing it will be about the same. Still I like the new pattern a lot better in terms of readability.

Reply:
I just wanted to avoid having to store the array length :( I will probably refactor the code to include @ImAllInNow's recommendation, and reuse the code between `add`/`remove` and `transfer`.